### DISCUSSION TITLE
Cant for the life of me get B800s working

### DISCUSSION THREAD

**Discussion by ZorkDaNerd** [QuantumEntangledAndy/neolink#252] [Q&A]
Hello, I have been beating my head against the wall for 2 days now. I have tried installing the original neolink repo, this repo, multiple versions of this repo. Different versions of gstream, docker containers, home assistant addon for neolink. None of them are allowing me to see the rtsp stream from the camera. I don't know what I am doing wrong and I feel like I have done everything right and it just isnt working.

If anyone could help me and run me through step by step with how they got the B800 to work and show up in something like Shodan I would be grateful!

**Comment by QuantumEntangledAndy** [comment]
Perhaps we can start with what we can do. Do you have any logs? It will tell me how far into the process you get, I.e. do we establish a connection to the camera.

**Reply by ZorkDaNerd** [reply]
This is the logs from running, it connects to the camera and sets up the streams but the RTSP IP is still 0.0.0.0 not something like 127.0.0.1 the instructions said that you can leave the bind at 0.0.0.0 and it will find an appropriate IP to put the camera stream on if I'm not mistaken. I then try to connect to that IP or if I set it to something like 127.0.0.1 and add it to Shodan or even just try to view the stream on VLC it will fail to grab it.
(I am redacting the IP but if the IP's have the same letters redacting them they are the same IP on my side.)
```
Running: /run.sh
--- VERSIONS ---
add-on version: 0.0.1
[2024-05-10T04:11:46Z INFO  neolink] Neolink ab947b895c3767ab404bfae71a5227a760117071 release
neolink version: neolink 0.6.3-rc.2
neolink mode: rtsp
neolink log: info
ATTENTION: if you expected a newer Neolink version, please reinstall this Add-on!
--- Neolink ---
[2024-05-10T04:11:46Z INFO  neolink] Neolink ab947b895c3767ab404bfae71a5227a760117071 release
[2024-05-10T04:11:46Z INFO  neolink::rtsp] Starting RTSP Server at 0.0.0.0:8554
[2024-05-10T04:11:46Z INFO  neolink::rtsp] B800: Rtsp Starting
[2024-05-10T04:11:46Z INFO  neolink::utils] B800: Connecting to camera at Address: X.X.X.X:9000
[2024-05-10T04:11:46Z INFO  neolink_core::bc_protocol] B800: Trying TCP discovery
[2024-05-10T04:11:46Z INFO  neolink_core::bc_protocol] B800: TCP Discovery success at X.X.X.X:9000
[2024-05-10T04:11:46Z INFO  neolink::utils] B800: Logging in
[2024-05-10T04:11:46Z INFO  neolink::utils] B800: Connected and logged in
[2024-05-10T04:11:48Z INFO  neolink::common::camthread] B800: Camera time is already set: 2024-05-09 20:11:48.0 +08:00:00
[2024-05-10T04:11:50Z INFO  neolink::common::neocam] B800: Model B800
[2024-05-10T04:11:50Z INFO  neolink::common::neocam] B800: Firmware Version v3.0.0.136_20120900
[2024-05-10T04:11:50Z INFO  neolink::rtsp::stream] B800: Available at /B800/main, /B800/Main, /B800/mainStream, /B800/MainStream, /B800/Mainstream, /B800/mainstream, /B800
[2024-05-10T04:12:06Z INFO  neolink::rtsp::stream] B800: Available at /B800/sub, /B800/Sub, /B800/subStream, /B800/SubStream, /B800/Substream, /B800/substream
```

This is what my config file looks like
```
bind = "0.0.0.0"

[[cameras]]
name = "B800"
username = "admin"
password = "Admin0000"
address = "X.X.X.X:9000"
```
Tell me if there is anything else I can provide to help diagnose it!
Thanks in advance for the help!

**Reply by QuantumEntangledAndy** [reply]
Not sure what shoudan is (quick Google suggests [this](https://en.m.wikipedia.org/wiki/Shodan_(website)). Are you aware of how IPs work. 127.0.0.1 exists ONLY on your machine. It cannot be reached externally at all. So if shoudan works as an external website it won't be able to reach it. You would need a routable public address, and all firewalls and NATs setup to allow an external website to access it. Perhaps try something simpler like vlc or ffmpeg first, these can work locally.

**Reply by QuantumEntangledAndy** [reply]
Just noticing from your logs your using the docker image from HA. Docker add a whole extra layer of complication as it makes 127.0.0.1 point to the INTERNAL docker node, not the host (unless it's in host mode). I assume you are exposing the ports properly in docker in this case. You might need to use your network IP address to connect

**Reply by QuantumEntangledAndy** [reply]
Since we seem to be in different time zones this could take awhile to resolve. So if possible can you share your setup. What you are running and how (I.e. what are your dockers, how are those dockers setup, what machine is neolink running on what machine is any other software on etc)

**Reply by ZorkDaNerd** [reply]
Sorry the software I was talking about was [Shinobi](https://docs.shinobi.video/). Correct I am newer to Networking and how it all works. The HA add-on was more for me to sanity check that it could see the camera. When I originally tried installing Neolink on a VM I was getting a Gstream failed to create object error. I just reated a new Ubuntu 24.04 VM in my proxmox server. I installed all of the Gstream packages and the libssl package. (It also cant be installed with just libssl, it needs something like -dev at the end unless I am misunderstanding [Link](https://packages.ubuntu.com/search?keywords=libssl))
Downloaded the latest ubuntu zip and unpacked it. I have the same config file as last time. Started it and this was the ouitput:
```
root@ubuntu:~/neolink_linux_x86_64_ubuntu# ./neolink rtsp --config=neolink.toml
[2024-05-10T20:22:02Z INFO  neolink] Neolink 459c9255ada7ebda281eba32a8a9b94a30c471f0 release
[2024-05-10T20:22:03Z INFO  neolink::rtsp] Starting RTSP Server at 0.0.0.0:8554
[2024-05-10T20:22:03Z INFO  neolink::rtsp] B800: Rtsp Staring
[2024-05-10T20:22:03Z INFO  neolink::utils] B800: Connecting to camera at Address: 192.168.50.182
[2024-05-10T20:22:03Z INFO  neolink_core::bc_protocol] B800: Trying TCP discovery
[2024-05-10T20:22:03Z INFO  neolink_core::bc_protocol] B800: TCP Discovery success at 192.168.50.182:9000
[2024-05-10T20:22:03Z INFO  neolink::utils] B800: Logging in
[2024-05-10T20:22:03Z INFO  neolink::utils] B800: Connected and logged in
[2024-05-10T20:22:05Z INFO  neolink::common::camthread] B800: Camera time is already set: 2024-05-10 12:22:05.0 +08:00:00
[2024-05-10T20:22:07Z WARN  neolink::common::pushnoti] Unable to save push notification details (fcm_token = "dRm2UqSJGjc:APA91bHChOhwYSJuaUlPozAnGgimN1JuTm-3a3C2vjTwzCn82uKRzQW2BAWuxTa5uKNH55rIHbA9rHzb1WPPQFqPDUNel3s4DVVgulancirTshm0VLauvniXBdwcNDgIQvWhyBBN3JEM"
    
    [gcm]
    android_id = "5359337628907994626"
    security_token = "6952217252728292582"
    
    [keys]
    public_key = "BEy2ZZit0fJf39GXEgdSAgjPhW605vc6hAixBUMIL_2nDICqcHqzkGyl8qAoNl1yT4lLhecaH6MiXA3RmWQJS5A"
    private_key = "ay0CVS8kCAtMKZPn0mbhn7P2YSiwludAdnvBJ4yBFmU"
    auth_secret = "12GFjUbNVgoF7ws73qBujw"
    ) to Some(
        "/root/.config/./neolink_token.toml",
    ) because of the error Os {
        code: 2,
        kind: NotFound,
        message: "No such file or directory",
    }
[2024-05-10T20:22:07Z INFO  neolink::common::neocam] B800: Model B800
[2024-05-10T20:22:07Z INFO  neolink::common::neocam] B800: Firmware Version v3.0.0.136_20120900
[2024-05-10T20:22:07Z INFO  neolink::rtsp::stream] B800: Avaliable at /B800/main, /B800/Main, /B800/mainStream, /B800/MainStream, /B800/Mainstream, /B800/mainstream, /B800
[2024-05-10T20:22:08Z INFO  neolink::rtsp::stream] B800: Avaliable at /B800/sub, /B800/Sub, /B800/subStream, /B800/SubStream, /B800/Substream, /B800/substream
```
I then installed [Shinobi](https://docs.shinobi.video/) and went through all of its step.
Neolink and Shinobi are both on the same VM. I first probed the rtsp://0.0.0.0:8554/B800 and Shinobi gave me this output:

```
streams :
    0 :
		index : 0
		codec_name : hevc
		codec_long_name : H.265 / HEVC (High Efficiency Video Coding)
		profile : Main
		codec_type : video
		codec_tag_string : [0][0][0][0]
		codec_tag : 0x0000
		width : 3840
		height : 2160
		coded_width : 3840
		coded_height : 2160
		closed_captions : 0
		film_grain : 0
		has_b_frames : 0
		pix_fmt : yuvj420p
		level : 150
		color_range : pc
		color_space : bt709
		color_transfer : bt709
		color_primaries : bt709
		chroma_location : left
		refs : 1
		r_frame_rate : 20/1
		avg_frame_rate : 0/0
		time_base : 1/90000
		extradata_size : 88
		disposition :
			default : 0
			dub : 0
			original : 0
			comment : 0
			lyrics : 0
			karaoke : 0
			forced : 0
			hearing_impaired : 0
			visual_impaired : 0
			clean_effects : 0
			attached_pic : 0
			timed_thumbnails : 0
			non_diegetic : 0
			captions : 0
			descriptions : 0
			metadata : 0
			dependent : 0
			still_image : 0
   1 :
	index : 1
	codec_name : pcm_s16be
	codec_long_name : PCM signed 16-bit big-endian
	codec_type : audio
	codec_tag_string : [0][0][0][0]
	codec_tag : 0x0000
	sample_fmt : s16	
	sample_rate : 16000
	channels : 2
	channel_layout : stereo
	bits_per_sample : 16
	initial_padding : 0
	r_frame_rate : 0/0
	avg_frame_rate : 0/0
	time_base : 1/16000
	start_pts : 0
	start_time : 0.000000
	bit_rate : 512000
	disposition :
		default : 0
		dub : 0
		original : 0
		comment : 0
		lyrics : 0
		karaoke : 0
		forced : 0
		hearing_impaired : 0
		visual_impaired : 0
		clean_effects : 0
		attached_pic : 0
		timed_thumbnails : 0
		non_diegetic : 0
		captions : 0
		descriptions : 0
		metadata : 0
		dependent : 0
		still_image : 0
format :
	filename : rtsp://0.0.0.0:8554/B800
	nb_streams : 2
	nb_programs : 0
	format_name : rtsp
	format_long_name : RTSP input
	start_time : 0.000000
	bit_rate : 512000
	probe_score : 100
	tags :
		title : Session streamed with GStreamer
		comment : rtsp-server
	url : rtsp://0.0.0.0:8554/B800
END
streams :
	0 :
		index : 0
		codec_name : hevc
		codec_long_name : H.265 / HEVC (High Efficiency Video Coding)
		profile : Main
		codec_type : video
		codec_tag_string : [0][0][0][0]
		codec_tag : 0x0000
		width : 3840
		height : 2160
		coded_width : 3840
		coded_height : 2160
		closed_captions : 0
		film_grain : 0
		has_b_frames : 0
		pix_fmt : yuvj420p
		level : 150
		color_range : pc
		color_space : bt709
		color_transfer : bt709
		color_primaries : bt709
		chroma_location : left
		refs : 1
		r_frame_rate : 20/1
		avg_frame_rate : 0/0
		time_base : 1/90000
		extradata_size : 88
		disposition :
			default : 0
			dub : 0
			original : 0
			comment : 0
			lyrics : 0
			karaoke : 0
			forced : 0
			hearing_impaired : 0
			visual_impaired : 0
			clean_effects : 0
			attached_pic : 0
			timed_thumbnails : 0
			non_diegetic : 0
			captions : 0
			descriptions : 0
			metadata : 0
			dependent : 0
			still_image : 0
	1 :
		index : 1
		codec_name : pcm_s16be
		codec_long_name : PCM signed 16-bit big-endian
		codec_type : audio
		codec_tag_string : [0][0][0][0]
		codec_tag : 0x0000
		sample_fmt : s16
		sample_rate : 16000
		channels : 2
		channel_layout : stereo
		bits_per_sample : 16
		initial_padding : 0
		r_frame_rate : 0/0
		avg_frame_rate : 0/0
		time_base : 1/16000
		start_pts : 0
		start_time : 0.000000
		bit_rate : 512000
		disposition :
			default : 0
			dub : 0
			original : 0
			comment : 0
			lyrics : 0
			karaoke : 0
			forced : 0
			hearing_impaired : 0
			visual_impaired : 0
			clean_effects : 0
			attached_pic : 0
			timed_thumbnails : 0
			non_diegetic : 0
			captions : 0
			descriptions : 0
			metadata : 0
			dependent : 0
			still_image : 0
	format :
		filename : rtsp://0.0.0.0:8554/B800
		nb_streams : 2
		nb_programs : 0
		format_name : rtsp
		format_long_name : RTSP input
		start_time : 0.000000
		bit_rate : 512000
		probe_score : 100
		tags :
			title : Session streamed with GStreamer
			comment : rtsp-server
	url : rtsp://0.0.0.0:8554/B800
END
```

I tried to add the stream to Shinobi with rtsp://0.0.0.0:8554/B800. But the stream window constantly restarts and just shows a black box. I also tried the RTSP with /sub /main and when I probed them the output from Shinobi looked very similar. But when I try to add the stream it just starts and dies.
I am only needing the streams to be visible to this VM since I will be using the security camera software to look at and record the camera streams.
This is what I am running the vm with this setup:
![image](https://github.com/QuantumEntangledAndy/neolink/assets/103788783/a4ad3aa3-8d9d-41f7-bbac-be5fb5f2780e)
I don't know if I need to do anything with ports on this VM, that stuff I am still a bit fuzzy with how it all works.
I also don't know if this may be more of an error with Shinobi or not.
Thanks again!

I can also install frigate to see if that software would play better.

**Reply by QuantumEntangledAndy** [reply]
If they in the same vm (not separate containers or anything). Then you can use 127.0.0.1. The address 0.0.0.0 will never work for a connection. It means all available network devices, which you can use when listening but not when connecting. By putting neolink config to 0.0.0.0 we are telling neolink to listen for a connection of all network devices. If you run `ip addr` you should see all your devices and their IPs, it will listen on all of those. When connecting you can only connect to one of them (so 0.0.0.0 will fail from the client). Instead you need to use a real IP address from the list shown in `ip addr`. There is a special one always a available on what's called the loop back 127.0.0.1. You can use the loop back to connect if they are on the same machine.

**Reply by QuantumEntangledAndy** [reply]
Anyways it looks like shinobi is translating 0.0.0.0 into a real ip for you automatically since we got some connection details when you used it.

Perhaps we can try some setting changes on neolink end. The software is not perfect and has to deal with translating Reolink format which can make it difficult to pipe though very HD connections. We will reduce things down to a minimum (then add them back later). First add `stream = "sub"` to the `[[cameras]]` section (this will only creat the sub stream which may help as some cameras don't like multiple connections at once). Only add one camera to neolink for now (I think you only have one but want to check)

Please try to connect from the terminal with `ffmpeg -i rtsp://127.0.0.1:8554/B800/sub test.mkv` it will create a mkv which you can copy to another machine to watch and check.
